src/OpenLoco/src/GameCommands/Road/RemoveRoadStation.cpp

```diff
@@ -0,0 +1,156 @@
+            }
+            if (elRoad->baseHeight() != pos.z)
+            {
+                continue;
+            }
+            if (elRoad->unkDirection() != rotation)
+            {
+                continue;
+            }
+            if (elRoad->roadId() != roadId)
+            {
+                continue;
+            }
+            // TODO: not seeing this check in disasm
+            if (elRoad->roadObjectId() != roadObjectId)
+            {
+                continue;
+            }
+            if (elRoad->sequenceIndex() != index)
+            {
+                continue;
+            }
+            return elRoad;
+        }
+        return nullptr;
+    }
+
+    // 0x0048D2AC
+    static currency32_t removeRoadStation(const RoadStationRemovalArgs& args, const uint8_t flags)
+    {
+        setExpenditureType(ExpenditureType::Construction);
+        setPosition(args.pos + World::Pos3(16, 16, 0));
+        bool updateStationTileRegistration = true;
+
+        // Get road at position
+        auto* initialElRoad = getElRoad(args.pos, args.rotation, args.type, args.roadId, args.index);
+        if (initialElRoad == nullptr)
+        {
+            return FAILURE;
+        }
+
+        // No station element on this road?
+        if (!initialElRoad->hasStationElement())
+        {
+            return FAILURE;
+        }
+
+        // Find the station element for an ownership check
+        {
+            World::StationElement* stationEl = nullptr;
+            while (auto* nextElement = initialElRoad->next())
+            {
+                stationEl = nextElement->as<World::StationElement>();
+                if (stationEl != nullptr)
+                {
+                    break;
+                }
+
+                if (nextElement->isLast())
+                {
+                    return FAILURE;
+                }
+            }
+
+            // NB: vanilla would query owner from station object, not the station element
+            if (!sub_431E6A(stationEl->owner(), reinterpret_cast<const World::TileElement*>(stationEl)))
+            {
+                return FAILURE;
+            }
+        }
+
+        const auto& roadPieces = World::TrackData::getRoadPiece(args.roadId);
+        const auto& argPiece = roadPieces[args.index];
+        const auto roadStart = args.pos - World::Pos3(Math::Vector::rotate(World::Pos2(argPiece.x, argPiece.y), args.rotation), argPiece.z);
+
+        StationId foundStationId = StationId::null;
+        currency32_t totalCost = 0;
+
+        for (auto& piece : roadPieces)
+        {
+            const auto roadLoc = roadStart + World::Pos3{ Math::Vector::rotate(World::Pos2{ piece.x, piece.y }, args.rotation), piece.z };
+
+            auto* elRoad = getElRoad(roadLoc, args.rotation, args.type, args.roadId, piece.index);
+
+            if (elRoad != nullptr)
+            {
+                auto* nextEl = elRoad->next();
+                auto* stationEl = nextEl->as<World::StationElement>();
```

Comment:
this isn't correct vanilla does the check to find the last road element (plus height check) and the station will be after it.

Reply:
ahh actually its doing this (pinched from create)
```cpp

                elRoads[0] = nullptr;
                elRoads[1] = nullptr;
                auto tile = World::TileManager::get(roadLoc);
                for (auto& el : tile)
                {
                    auto* elRoad = el.as<World::RoadElement>();
                    if (elRoad == nullptr)
                    {
                        continue;
                    }
                    if (elRoad->baseHeight() != roadLoc.z)
                    {
                        continue;
                    }
                    elRoads[1] = elRoad;
                    if (elRoads[0] == nullptr)
                    {
                        elRoads[0] = elRoad;
                    }
                }
```
It gets the begin and end pointers of the road elements.